### PR TITLE
[Gluten-986] Support task input metrics, fix wrong output metrics.

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

Current `MetricsUpdater::updateOutputMetrics` is unnecessary and has bug, it will count twice output metrics.

image

Implement task input metrics, only Scan operator could update that.

image

### CODE REVIEW

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala [hunk 1]
```diff
@@ -158,6 +159,7 @@ class CHIteratorApi extends IIteratorApi with Logging with LogLevelUtil {
       context: TaskContext,
       pipelineTime: SQLMetric,
       updateOutputMetrics: (Long, Long) => Unit,
```
Comment (rui-mo) [inline backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:161]: Can this be removed if it is not needed?

File: gluten-core/src/main/scala/io/glutenproject/execution/CoalesceExecTransformer.scala [hunk 2]
```diff
@@ -47,8 +47,11 @@ case class CoalesceExecTransformer(numPartitions: Int, child: SparkPlan)
     throw new UnsupportedOperationException(s"This operator doesn't support getBuildPlans.")
   }
 
-  override def getStreamedLeafPlan: SparkPlan = {
-    throw new UnsupportedOperationException(s"This operator doesn't support getStreamedLeafPlan.")
+  override def getStreamedLeafPlan: SparkPlan = child match {
+    case c: TransformSupport =>
+      c.getStreamedLeafPlan
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/CoalesceExecTransformer.scala:52]: Why do we need this change?

**Review by rui-mo — COMMENTED** [review]
Hi @Yohahaha, for the output metrics issue, previously we found the `ouputRows` and `outputVectors` metrics from some Velox nodes are zero. That's why we use Scala side `updateOutputMetrics` to compensate for that. It would be great if we can make sure all the affected operators work well on the two output metrics after removal.

**Comment by Yohahaha** [comment]
> Hi @Yohahaha, for the output metrics issue, previously we found the `ouputRows` and `outputVectors` metrics from some Velox nodes are zero. That's why we use Scala side `updateOutputMetrics` to compensate for that. It would be great if we can make sure all the affected operators work well now on the two output metrics after removal.

I would follow this problem if it is happened again.

**Comment by rui-mo** [comment]
> I would follow this problem if it is happened again.

Got it. I agree to remove this workaround, and if we find the output metrics being 0, we need to check that issue in Velox.

**Review by FelixYBW — APPROVED** [review]

**Comment by Yohahaha** [comment]
Hi @rui-mo, could you help review again?

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala [hunk 3]
```diff
@@ -179,7 +180,6 @@ class CHIteratorApi extends IIteratorApi with Logging with LogLevelUtil {
 
       override def next(): Any = {
         val cb = resIter.next()
-        updateOutputMetrics(1, cb.numRows())
```
Comment (rui-mo) [inline backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:182]: Hi @zzcclp, is this function call still needed on CH backend?

**Comment by rui-mo** [comment]
> Hi @rui-mo, could you help review again?

@Yohahaha Thanks for your work. This PR LGTM on Velox backend, but I'm not sure if it is all right for CH backend. Maybe @zzcclp can help check that.

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala [hunk 4]
```diff
@@ -157,7 +158,7 @@ class CHIteratorApi extends IIteratorApi with Logging with LogLevelUtil {
       outputAttributes: Seq[Attribute],
       context: TaskContext,
       pipelineTime: SQLMetric,
-      updateOutputMetrics: (Long, Long) => Unit,
+      updateInputMetrics: (InputMetricsWrapper) => Unit,
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:161]: > @zzcclp we currently use the same metrics for Velox and Clickhouse. But it definitely isn't possible. Any good way to seperate these?

I will raise a pr to seperate these metrics for velox and ch backends in the next week.

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala (hunk 3 above)
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:182]: remove this first, if there is any issue, I will fix it later.

**Pull request merged**